### PR TITLE
Basic logging for HiveRegistry struct

### PR DESCRIPTION
**Pull request by kamilkisiela**

### CODE REVIEW

File: packages/libraries/router/src/registry_logger.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
+use std::{env, fmt};
+
+static LOG_LEVEL_NAMES: [&str; 5] = ["ERROR", "WARN", "INFO", "DEBUG", "TRACE"];
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/registry_logger.rs:3]: maybe we can remove this bc it isn't used anywhere except in `Level::from_str`, and it just creates this extra jump to reveal the implementation of the method.

Comment (kamilkisiela) [reply packages/libraries/router/src/registry_logger.rs:3]: we need it for `as_str`

File: packages/libraries/router/src/registry_logger.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            println!("WARNING: {}", message);
+        }
+    }
+
+    pub fn error(&self, message: &str) {
+        if self.should_log(Level::Error) {
+            println!("ERROR: {}", message);
+        }
+    }
+}
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/registry_logger.rs:98]: maybe `LogLevel` is a better descriptive name.

```suggestion
#[repr(usize)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Debug)]
pub enum LogLevel {
    Error,
    Warn,
    Info,
    Debug,
    Trace,
}

impl LogLevel {
    fn from_usize(u: usize) -> Option<LogLevel> {
        match u {
            0 => Some(LogLevel::Error),
            1 => Some(LogLevel::Warn),
            2 => Some(LogLevel::Info),
            3 => Some(LogLevel::Debug),
            4 => Some(LogLevel::Trace),
            _ => None,
        }
    }

    fn from_str(s: &str) -> LogLevel {
        LOG_LEVEL_NAMES
            .iter()
            .position(|&name| name.eq_ignore_ascii_case(s))
            .map(|p| LogLevel::from_usize(p).expect("Hive failed to read the log level"))
            .expect("Hive failed to parse the log level filter")
    }

    pub fn as_str(&self) -> &'static str {
        LOG_LEVEL_NAMES[*self as usize]
    }
}

impl fmt::Display for Level {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.pad(self.as_str())
    }
}

#[derive(Clone, Debug)]
pub struct Logger {
    max_level: Level,
}

impl Logger {
    pub fn new() -> Logger {
        Self {
            max_level: LogLevel::from_str(
                env::var("HIVE_REGISTRY_LOG")
                    .unwrap_or_else(|_| "info".to_string())
                    .as_str(),
            ),
        }
    }

    fn should_log(&self, level: LogLevel) -> bool {
        self.max_level >= level
    }

    #[allow(dead_code)]
    pub fn trace(&self, message: &str) {
        if self.should_log(LogLevel::Trace) {
            println!("TRACE: {}", message);
        }
    }

    #[allow(dead_code)]
    pub fn debug(&self, message: &str) {
        if self.should_log(LogLevel::Debug) {
            println!("DEBUG: {}", message);
        }
    }

    pub fn info(&self, message: &str) {
        if self.should_log(LogLevel::Info) {
            println!("INFO: {}", message);
        }
    }

    #[allow(dead_code)]
    pub fn warn(&self, message: &str) {
        if self.should_log(LogLevel::Warn) {
            println!("WARNING: {}", message);
        }
    }

    pub fn error(&self, message: &str) {
        if self.should_log(LogLevel::Error) {
            println!("ERROR: {}", message);
        }
    }
}
```

File: packages/libraries/router/src/registry_logger.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
+use std::{env, fmt};
+
+static LOG_LEVEL_NAMES: [&str; 5] = ["ERROR", "WARN", "INFO", "DEBUG", "TRACE"];
+
+#[repr(usize)]
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/registry_logger.rs:5]: I don't think this attribute is needed
```suggestion
```

Comment (kamilkisiela) [reply packages/libraries/router/src/registry_logger.rs:5]: I think it's needed for a `>=` comparison, no?

Comment (YassinEldeeb) [reply packages/libraries/router/src/registry_logger.rs:5]: oh yes! I've just discovered that down

File: packages/libraries/router/src/registry_logger.rs (hunk 1 above)
Comment (YassinEldeeb) [reply packages/libraries/router/src/registry_logger.rs:3]: just a suggestion, both works perfectly!

File: packages/libraries/router/src/registry_logger.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+impl fmt::Display for Level {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.pad(self.as_str())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Logger {
+    max_level: Level,
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/registry_logger.rs:48]:
```suggestion
    max_level: LogLevel,
```

File: packages/libraries/router/src/registry_logger.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    pub fn as_str(&self) -> &'static str {
+        LOG_LEVEL_NAMES[*self as usize]
+    }
+}
+
+impl fmt::Display for Level {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.pad(self.as_str())
+    }
+}
```
[the remark this answers is not in the document]
Comment (YassinEldeeb) [reply packages/libraries/router/src/registry_logger.rs:44]: if we need it, then we should change the `Level` as well.

```suggestion
impl fmt::Display for LogLevel {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.pad(self.as_str())
    }
}
```

File: packages/libraries/router/src/registry_logger.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        }
+    }
+
+    fn from_str(s: &str) -> LogLevel {
+        LOG_LEVEL_NAMES
+            .iter()
+            .position(|&name| name.eq_ignore_ascii_case(s))
+            .map(|p| LogLevel::from_usize(p).expect("Hive failed to read the log level"))
+            .expect("Hive failed to parse the log level filter")
+    }
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/registry_logger.rs:33]: I think here there's a possibility to eliminate an unnecessary panic call, and make it more graceful to failure

```suggestion
    fn from_str(s: &str) -> Option<LogLevel> {
        match s {
            "error" => Some(LogLevel::Error),
            "warn" => Some(LogLevel::Warn),
            "info" => Some(LogLevel::Info),
            "debug" => Some(LogLevel::Debug),
            "trace" => Some(LogLevel::Trace),
            _ => None,
        }
    }
```

Comment (YassinEldeeb) [reply packages/libraries/router/src/registry_logger.rs:33]: but we'll also need to modify the `from_str` call down below if you accepted it.

File: packages/libraries/router/src/registry_logger.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+impl Logger {
+    pub fn new() -> Logger {
+        Self {
+            max_level: LogLevel::from_str(
+                env::var("HIVE_REGISTRY_LOG")
+                    .unwrap_or_else(|_| "info".to_string())
+                    .as_str(),
+            ),
+        }
```
Comment (YassinEldeeb) [inline packages/libraries/router/src/registry_logger.rs:59]: only if you accepted the suggestion of updating the `from_str`method
 
```suggestion
        Self {
            max_level: LogLevel::from_str(
                &env::var("HIVE_REGISTRY_LOG").unwrap_or_else(|_| "info".to_string()),
            )
            .unwrap_or(LogLevel::Info),
        }
```

**Review by YassinEldeeb — APPROVED** [review]

**Pull request merged**